File: python/hamming/hamming.py
"""
Module for calculating the Hamming distance between two strings.
"""


def distance(strand_a: str, strand_b: str) -> int:
    """
    Calculates the Hamming distance between two strings.

    The Hamming distance is the number of positions at which the strings differ.
    The strings must be of equal length.

    Args:
        strand_a (str): The first string.
        strand_b (str): The second string.

    Returns:
        int: The Hamming distance between the two strings.

    Raises:
        ValueError: If the strings have different lengths.
    """
    if len(strand_a) != len(strand_b):
        raise ValueError("Strands must be of equal length.")
    return sum(1 for pair in zip(strand_a, strand_b) if pair[0] != pair[1])

File: csharp/hamming/Hamming.cs
/// <summary>
/// A static class to calculate the Hamming distance between two strings.
/// </summary>
public static class Hamming
{
    /// <summary>
    /// Calculates the Hamming distance between two DNA strands.
    /// </summary>
    /// <remarks>
    /// The Hamming distance is the number of positions at which the corresponding symbols are different.
    /// This method is only defined for strands of equal length.
    /// </remarks>
    /// <param name="firstStrand">The first DNA strand.</param>
    /// <param name="secondStrand">The second DNA strand.</param>
    /// <returns>The Hamming distance between the two strands.</returns>
    /// <exception cref="ArgumentException">Thrown when the strands have different lengths.</exception>
    public static int Distance(string firstStrand, string secondStrand)
    {
        const string errorMessage = "It is not possible to calculate the clamping distance for ropes of different sizes.";
        if (firstStrand.Length != secondStrand.Length)
        {
            throw new ArgumentException(errorMessage);
        }
        int hamming = 0;

        for (int i = 0; i < firstStrand.Length; i++)
        {
            if (firstStrand[i] != secondStrand[i])
            {
                hamming++;
            }
        }

        return hamming;
    }
}